File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.cc
#include <bits/stdc++.h>
using namespace std;

const int MAX = 10001;
int N, M, a, b, mx, dp[MAX];
bool visited[MAX];
vector<int> ret, adj[MAX];

int go(int here) {
    int cnt = 1;
    visited[here] = 1;

    for (int there: adj[here]) {
        if (visited[there]) continue;
        cnt += go(there);
    }
    return cnt;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> N >> M;
    for (int i=0; i<M; i++) {
        cin >> a >> b;
        adj[b].push_back(a);
    }

    for (int i=1; i<=N; i++) {
        memset(visited, 0, sizeof(visited));
        dp[i] = go(i);
        mx = max(mx, dp[i]);
    }

    for (int i=1; i<=N; i++) {
        if (mx == dp[i]) cout << i << ' ';
    }
    return 0;
}

File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.py
import sys
from collections import deque
input = lambda: sys.stdin.readline().rstrip()

MAX = 10001
adj = [[] for _ in range(MAX)]

def bfs(here):
    global visit_id

    visit_id += 1
    q = deque([here])
    visited[here] = visit_id
    cnt = 1

    while q:
        here = q.popleft()
        for there in adj[here]:
            if visited[there] == visit_id: 
                continue
            visited[there] = visit_id
            q.append(there)
            cnt += 1

    return cnt


if __name__ == "__main__":
    N, M = map(int, input().split())
    for _ in range(M):
        a, b = map(int, input().split())
        adj[b].append(a)

    mx = 0
    dp = [0] * MAX
    visited = [0] * MAX
    visit_id = 0
    for i in range(1, N+1):
        dp[i] = bfs(i)
        mx = max(mx, dp[i])

    for i in range(1, N+1):
        if dp[i] == mx:
            print(i, end=' ')
